Tree-sitter: rewrite update script partially in python

Profpatsch opened the pull request:
###### Description of changes

The update script is getting out of hand, so we should rewrite it in
something that resembles sanity, i.e. Python.

It does not go all the way (yet), mainly because file wrangling & parallel execution was easier on the nix/bash side. But the fetching and (most of) the json wrangling is done in python now instead of bash/jq.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Profpatsch committed:
tree-sitter: partially rewrite update script in python

The update script is getting out of hand, so we should rewrite it in
something that resembles sanity, i.e. Python

This is the first step, rewriting the part that checks and fetches a
release.

Next up is integrating latestGithubRepos into that script.

Profpatsch committed:
tree-sitter/update: prepare moving more stuff to python

I don’t want to split up the python code, so for now I’m just passing
a “mode” to the script to run a different subset of the functionality.

Profpatsch committed:
tree-sitter/update: factor out github url

Profpatsch committed:
tree-sitter/update: move pyhon impl into its own file

This helps with syntax highlighting & editor support

Profpatsch committed:
tree-sitter/update: get executables from nix instead of environment

This was a small bug, the update script should of course not depend on
any executables in the environment.

Since we don’t use a nix string anymore, interpolation is out of the
question, so we pass it as arglib environment json string.

Profpatsch committed:
tree-sitter/update: Fetch the existing repos from python as well

I also used the chance to add some type annotations and make sure they
get accepted by mypy.

Profpatsch committed:
tree-sitter/update: Write files atomically

Otherwise you can’t interrupt the process without creating
broken/half-written files.

Profpatsch committed:
tree-sitter/update: Fetch repositories in parallel

This was the main annoyance with the existing script, it would fetch
one repository after the other instead of multiple in parallel.
The `xe` tool is a simple `xargs`-like that can do that.

We put the json information for all repositories into a file with one
object per line, then pass that to xe.

Profpatsch committed:
tree-sitter: Update grammars

Profpatsch commented:
@oxalica @Hoverbear @Kranzes do you want to run the script on your machines to see whether it still works as before?

Profpatsch commented:
If nobody else wants to test it, I’d merge tomorrow or so

oxalica commented on pkgs/development/tools/parsing/tree-sitter/update.nix:
```diff
@@ -392,89 +395,83 @@ let
   # TODO
   urlEscape = x: x;
```
Unused.

oxalica reviewed (changes requested):
> It does not go all the way (yet), mainly because file wrangling & parallel execution was easier on the nix/bash side. But the fetching and (most of) the json wrangling is done in python now instead of bash/jq.

Mixing python and bash scripts seems messy to me. I'd prefer rewrite the whole `update.nix` in python and not passing data around.

Some notes:
- Builtin `urllib.request` for fetching. It should not be more complex than spawning `curl`.
- Builtin `threading` for concurrency.
- No `async` needed I think.

Profpatsch commented:
@oxalica In the first step I wanted to keep the semantics the same and only get rid of the bash array code & duplication. Sure it could be moved all the way, thought I’m not really sure using `threading` in python is a good idea. `xe` handles this pretty well, and e.g. also does stuff like forwarding SIGINT correctly.

oxalica commented:
> In the first step I wanted to keep the semantics the same and only get rid of the bash array code & duplication.

IMO the current change does NOT simplify things but adding more complexity. Especially the cross-language json passing in `fetchImpl`, multi-interpreter `forEachParallel` and actually-two-lines-of-python `atomically-write`.

I'd rather keep the bash script as-is, or have a mature-and-complete python or whatever script rewrite. Mixing three languages for an update script is surely a bad idea. 

> Sure it could be moved all the way, thought I’m not really sure using `threading` in python is a good idea. `xe` handles this pretty well, and e.g. also does stuff like forwarding SIGINT correctly.

You are worrying about the signal issue because you are doing multi-processing in bash. If everything is done in one process (eg, one python script without spawning), SIGINT simply stops it and nothing weird would happen.

BTW: The style config [pkgs/development/tools/parsing/tree-sitter/mypy.ini](https://github.com/NixOS/nixpkgs/pull/190083/files#diff-7225638ac7db2fab0adabb3f558c29beeb0037aedfbeed0924d520989a217e43) should not be included in tree.

Profpatsch committed:
tree-sitter/update: ARGLIB_JSON -> BINARIES, factor out wrapper

arglib is a term I coined for passing arguments as structured
environment variable, but in this case it’s probably easier to
understand if we just call it `BINARIES`.

Factors out the code that passes the environment to another script.

teto commented on pkgs/development/tools/parsing/tree-sitter/update_impl.py:
```diff
@@ -0,0 +1,133 @@
+import sys
+from typing import Generator, Any, Literal
+
+debug: bool = True if os.environ.get("DEBUG", False) else False
```
I like the usage of argparse: it allows to self document the usage of the script.

teto commented:
I wish we could have a "unified" interface between the different package updater. If the script could accept the same arguments as  pkgs/applications/editors/vim/plugins/update.py  / maintainers/scripts/pluginupdate.py e.g., `--github-token` , `--proc` , `--no-commit` and so on.

Profpatsch committed:
tree-sitter/update: Move json file output to python

Removes the second-to-last use of jq from the shell script.

Profpatsch commented:
> IMO the current change does NOT simplify things but adding more complexity.

yes of course, that’s the whole idea behind incremental refactoring. First the complexity increases, then it shrinks.

> and actually-two-lines-of-python `atomically-write`.

I’d like to switch to a two lines python implementation, but first all uses have to move to the python script.



> BTW: The style config [pkgs/development/tools/parsing/tree-sitter/mypy.ini](https://github.com/NixOS/nixpkgs/pull/190083/files#diff-7225638ac7db2fab0adabb3f558c29beeb0037aedfbeed0924d520989a217e43) should not be included in tree.

Why is that?

Profpatsch commented:
> (eg, one python script without spawning)

That’s a hefty assumption though. How would you replace shelling `nix-prefetch-git` in the python script?

oxalica commented:
> > (eg, one python script without spawning)
> 
> That’s a hefty assumption though. How would you replace shelling `nix-prefetch-git` in the python script?

Oh, that's valid. The main issue of current impl is about mixing logic and data in bash/nix/python parts and cross-calling around. `curl` and `nix-prefetch` invocations are acceptable though.

> yes of course, that’s the whole idea behind incremental refactoring. First the complexity increases, then it shrinks.

You can mark this PR as draft if you are not complete.

Profpatsch commented:
I was hoping to merge the partially-migrated version asap, since I think it already adds value (people can use the python code if any changes need to be made).
I will migrate the rest as far as possible in the future, but I can’t promise it will be soon.

The final version will always involve passing executables via the environment, because the alternative is writing the script as faux-python with `@interpolations@`, which I think is a non-solution.

The only thing that will probably go away is the remaining bash code in `checkTreeSitterRepos`, `update-all-grammars` and `atomically-write`.

SuperSandro2000 commented:
> I will migrate the rest as far as possible in the future, but I can’t promise it will be soon.

IMO then I would rather keep the current version and do a full conversion in one go to not have an intermediate very complex state like @oxalica already suggested.

Profpatsch commented:
@SuperSandro2000 I wasn’t aware you were maintaining the tree-sitter closure?

SuperSandro2000 commented:
I already added some parsers https://github.com/NixOS/nixpkgs/pull/177354

Profpatsch commented:
How would you have expected it to work?

Profpatsch committed:
tree-sitter/update: BINARIES -> ARGS.binaries

I want to pass more information in the next step, so binaries are just
one of multiple elements of the json value.

Generate a json file via `format` so that we get nicer
formatting (`lib.generators.toJSON` uses `builtins.toJSON` which does
not add any whitespace).

Profpatsch committed:
tree-sitter/update: move checkTreeSitterRepos into python impl

Direct translation of the jq set logic.

Profpatsch committed:
tree-sitter/update: move printing of the import nix file to python

This is a bit crappy, but the complexity should go away once it’s done
in all python. Might have been wiser to rewrite the other stuff first.

Profpatsch committed:
tree-sitter/update: move atomically-write to python

It should do the same thing, but take with a grain of salt maybe.

Profpatsch committed:
tree-sitter/update: rename fetchImpl to updateImpl

Profpatsch commented:
> How would you have expected it to work?

Serious question, maybe I’m missing some consensus on how update scripts should be written and/or integrated into the workflow.

Profpatsch committed:
tree-sitter/update: fetch orgas and directly check in python

Profpatsch commented:
Rebased, merging now so that further improvements can be based on the python script.
